### PR TITLE
planner/cascades: Add ImplementationRule for Window

### PR DESCRIPTION
**Pull request by edytagarbarz**
### What problem does this PR solve? 
Part of #13709 
Add ImplementationRule for Window

### CODE REVIEW

File: planner/implementation/simple_plans.go [hunk 1]
```diff
@@ -203,3 +203,19 @@ func (impl *MaxOneRowImpl) CalcCost(outCount float64, children ...memo.Implement
...
+	baseImpl
+}
+
+// NewWindowImpl creates a new WindowImpl.
+func NewWindowImpl(window *plannercore.PhysicalWindow) *WindowImpl {
+	return &WindowImpl{baseImpl{plan: window}}
+}
+
+// CalcCost implements Implementation CalcCost interface.
+func (impl *WindowImpl) CalcCost(outCount float64, children ...memo.Implementation) float64 {
```
Comment (edytagarbarz) [inline planner/implementation/simple_plans.go]: Consulted with @francis0407 . The Window cost is not included in the old planner and this is to be fixed later.

File: planner/cascades/implementation_rules.go [hunk 2]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+		PartitionBy:     lw.PartitionBy,
+		OrderBy:         lw.OrderBy,
+		Frame:           lw.Frame,
+	}.Init(
+		lw.SCtx(),
+		expr.Group.Prop.Stats,
+		lw.SelectBlockOffset(),
+		&property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: byItems, Enforced: true},
+	)
+	physicalWindow.SetSchema(lw.Schema())
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]: Use `expr.Group.Prop.Schema` or `expr.Schema()` here.

File: planner/cascades/implementation_rules.go [hunk 3]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+type ImplWindow struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (w *ImplWindow) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	lw := expr.ExprNode.(*plannercore.LogicalWindow)
+	var byItems []property.Item
+	byItems = append(byItems, lw.PartitionBy...)
+	byItems = append(byItems, lw.OrderBy...)
+	childProperty := &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: byItems, Enforced: true}
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]: `Enforced` is not used in cascades planner, you can just leave it.

File: planner/cascades/implementation_rules.go [hunk 4]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+	physicalWindow := plannercore.PhysicalWindow{
+		WindowFuncDescs: lw.WindowFuncDescs,
+		PartitionBy:     lw.PartitionBy,
+		OrderBy:         lw.OrderBy,
+		Frame:           lw.Frame,
+	}.Init(
+		lw.SCtx(),
+		expr.Group.Prop.Stats,
+		lw.SelectBlockOffset(),
+		&property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: byItems, Enforced: true},
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]: Ditto

File: planner/cascades/implementation_rules.go [hunk 5]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
 		&property.PhysicalProperty{ExpectedCnt: 2})
 	return impl.NewMaxOneRowImpl(physicalMaxOneRow), nil
 }
+
+// ImplWindow implements LogicalWindow to PhysicalWindow
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]:
```suggestion
// ImplWindow implements LogicalWindow to PhysicalWindow.
```

File: planner/cascades/implementation_rules.go [hunk 6]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+func (w *ImplWindow) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	lw := expr.ExprNode.(*plannercore.LogicalWindow)
+	var byItems []property.Item
+	byItems = append(byItems, lw.PartitionBy...)
+	byItems = append(byItems, lw.OrderBy...)
+	childProperty := &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: byItems, Enforced: true}
+	return prop.IsPrefix(childProperty)
+}
+
+// OnImplement implements ImplementationRule OnImplement interface
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]:
```suggestion
// OnImplement implements ImplementationRule OnImplement interface.
```

File: planner/cascades/implementation_rules.go [hunk 7]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+	byItems = append(byItems, lw.PartitionBy...)
+	byItems = append(byItems, lw.OrderBy...)
+	physicalWindow := plannercore.PhysicalWindow{
+		WindowFuncDescs: lw.WindowFuncDescs,
+		PartitionBy:     lw.PartitionBy,
+		OrderBy:         lw.OrderBy,
+		Frame:           lw.Frame,
+	}.Init(
+		lw.SCtx(),
+		expr.Group.Prop.Stats,
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]:
```suggestion
		expr.Group.Prop.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt)
```

File: planner/cascades/implementation_rules.go [hunk 8]
```diff
@@ -519,3 +522,38 @@ func (r *ImplMaxOneRow) OnImplement(expr *memo.GroupExpr, reqProp *property.Phys
...
+type ImplWindow struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (w *ImplWindow) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	lw := expr.ExprNode.(*plannercore.LogicalWindow)
+	var byItems []property.Item
+	byItems = append(byItems, lw.PartitionBy...)
+	byItems = append(byItems, lw.OrderBy...)
+	childProperty := &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64, Items: byItems}
```
Comment (francis0407) [inline planner/cascades/implementation_rules.go]: `ExpectedCnt` is also unused here, you can remove it.

**Pull request merged**